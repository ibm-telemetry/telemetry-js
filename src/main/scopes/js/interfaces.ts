/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { JsNodeHandler } from './node-handlers/js-node-handler.js'
import { NodeValueHandler } from './node-handlers/value-handlers/node-value-handler.js'

export interface JsImport {
  name: string
  path: string
  isDefault: boolean
  isAll: boolean
  rename?: string
}

type JsNodeHandlerClass<DataType> = new (
  node: ts.SourceFile,
  logger: Logger
) => JsNodeHandler<DataType>

export type JsNodeHandlerMap = Partial<Record<ts.SyntaxKind, JsNodeHandlerClass<unknown>>>

export type NodeValue = string | number | boolean | ComplexValue | null | undefined

type NodeValueHandlerProducer = new (node: ts.SourceFile, logger: Logger) => NodeValueHandler

export type NodeValueHandlerMap = Partial<Record<ts.SyntaxKind, NodeValueHandlerProducer>>
