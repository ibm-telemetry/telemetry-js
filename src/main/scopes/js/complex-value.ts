/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Object representing a complex value.
 */
export class ComplexValue {
  constructor(public readonly complexValue: unknown) {}
}
