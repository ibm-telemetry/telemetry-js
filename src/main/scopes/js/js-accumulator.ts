/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { JsImport } from '../js/interfaces.js'

/**
 * Base class for all JS Accumulators.
 * Responsible for maintaining an aggregated state of imports and other elements.
 */
export abstract class JsAccumulator {
  public imports: JsImport[]

  constructor() {
    this.imports = []
  }
}
