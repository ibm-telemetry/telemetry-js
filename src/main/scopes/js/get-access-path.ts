/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import type { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { getNodeValueHandler } from './node-value-handler-map.js'

/**
 * Constructs a JsToken object from a given Identifier type AST node.
 *
 * @param node - TS node to retrieve access path for.
 * @param sourceFile - Root AST node.
 * @param logger - A logger instance.
 * @param currAccessPath - For internal use only, tracks current constructed access path.
 * @param topLevel - For internal use only, tracks top level function call.
 * @returns Access path to node represented as string array (chunks).
 */
export default function getAccessPath(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  logger: Logger,
  currAccessPath: string[] = [],
  topLevel: boolean = true
): string[] {
  switch (node.kind) {
    case ts.SyntaxKind.Identifier:
      return [...currAccessPath, (node as ts.Identifier).escapedText.toString()]
    case ts.SyntaxKind.PropertyAccessExpression:
      currAccessPath.push((node as ts.PropertyAccessExpression).name.escapedText.toString())
      break
    case ts.SyntaxKind.ElementAccessExpression: {
      const argumentExpression = (node as ts.ElementAccessExpression).argumentExpression
      const data = getNodeValueHandler(argumentExpression.kind, sourceFile, logger).getData(
        argumentExpression
      )
      if (data instanceof ComplexValue) {
        currAccessPath = [
          ...currAccessPath,
          getAccessPath(argumentExpression, sourceFile, logger, [], false).reverse().join('.')
        ]
      } else {
        currAccessPath.push(data?.toString() ?? '')
      }
      break
    }
  }

  let accessPath = currAccessPath

  if ('expression' in node) {
    accessPath = getAccessPath(
      node.expression as ts.Node,
      sourceFile,
      logger,
      currAccessPath,
      false
    )
  }

  return topLevel ? accessPath.reverse() : accessPath
}
