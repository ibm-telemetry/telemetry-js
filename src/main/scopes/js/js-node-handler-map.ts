/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import { JsNodeHandlerMap } from '../js/interfaces.js'
import { ImportNodeHandler } from '../js/node-handlers/import-node-handler.js'
import { CallExpressionNodeHandler } from './node-handlers/tokens-and-functions-handlers/call-expression-node-handler.js'
import { ElementAccessExpressionNodeHandler } from './node-handlers/tokens-and-functions-handlers/element-access-expression-node-handler.js'
import { IdentifierNodeHandler } from './node-handlers/tokens-and-functions-handlers/identifier-node-handler.js'
import { PropertyAccessExpressionNodeHandler } from './node-handlers/tokens-and-functions-handlers/property-access-expression-node-handler.js'

/**
 * Maps node kinds to handlers that know how to process them to generate JsFunction and JsToken
 * metrics for the JsScope.
 */
export const jsNodeHandlerMap: JsNodeHandlerMap = {
  [ts.SyntaxKind.ImportDeclaration]: ImportNodeHandler,
  [ts.SyntaxKind.CallExpression]: CallExpressionNodeHandler,
  [ts.SyntaxKind.PropertyAccessExpression]: PropertyAccessExpressionNodeHandler,
  [ts.SyntaxKind.ElementAccessExpression]: ElementAccessExpressionNodeHandler,
  [ts.SyntaxKind.Identifier]: IdentifierNodeHandler
}
