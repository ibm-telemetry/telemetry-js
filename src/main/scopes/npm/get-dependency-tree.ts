/*
 * Copyright IBM Corp. 2024, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { DirectoryEnumerator } from '../../core/directory-enumerator.js'
import type { Logger } from '../../core/log/logger.js'
import { runCommand } from '../../core/run-command.js'
import { NoNodeModulesFoundError } from '../../exceptions/no-node-modules-found-error.js'
import { hasNodeModulesFolder } from './has-node-modules-folder.js'
import type { DependencyTree } from './interfaces.js'

const cache = new Map<string, Promise<DependencyTree>>()

/**
 * Obtains data relating to package info and dependencies recursively,
 * starting from the root-most package.
 *
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param logger - Logger instance.
 * @returns Object containing package and dependencies tree for the given root.
 */
export async function getDependencyTree(
  cwd: string,
  root: string,
  logger: Logger
): Promise<DependencyTree> {
  const cacheKey = `${cwd} ${root}`

  if (cache.has(cacheKey)) {
    logger.debug('getDependencyTree cache hit for ' + cacheKey)
    const data = await (cache.get(cacheKey) as Promise<DependencyTree>)
    logger.traceExit('', 'getDependencyTree', data)
    return data
  }

  const getNpmTree = async () => {
    const dirs = await new DirectoryEnumerator(logger).find(cwd, root, hasNodeModulesFolder)
    const topMostDir = dirs.pop()

    if (topMostDir === undefined) {
      throw new NoNodeModulesFoundError(cwd, root)
    }

    // Allow this command to try and obtain results even if it exited with a total or partial
    // error
    const commandResult = await runCommand(
      'npm ls --all --json',
      logger,
      { cwd: topMostDir },
      false
    )

    const npmTree = JSON.parse(commandResult.stdout)

    logger.traceExit('', 'getDependencyTree inner promise', npmTree)
    return npmTree
  }

  cache.set(cacheKey, getNpmTree())

  const data = await (cache.get(cacheKey) as Promise<DependencyTree>)

  return data
}
