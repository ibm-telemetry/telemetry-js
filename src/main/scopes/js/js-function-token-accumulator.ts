/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { JsFunction, JsImport, JsToken } from '../js/interfaces.js'
import { JsAccumulator } from '../js/js-accumulator.js'

/**
 * Responsible for maintaining an aggregated state of imports, functions and tokens.
 */
export class JsFunctionTokenAccumulator extends JsAccumulator {
  public readonly tokens: JsToken[]
  public readonly functions: JsFunction[]
  public readonly functionImports: Map<JsFunction, JsImport>
  public readonly tokenImports: Map<JsToken, JsImport>

  constructor() {
    super()
    this.tokens = []
    this.functions = []
    this.functionImports = new Map()
    this.tokenImports = new Map()
  }
}
