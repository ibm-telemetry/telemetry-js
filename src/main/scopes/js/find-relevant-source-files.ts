/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import getPropertyByPath = require('lodash/get.js')
import * as path from 'node:path'

import { ObjectPath } from 'object-scan'

import { getTrackedSourceFiles } from '../../core/get-tracked-source-files.js'
import { Logger } from '../../core/log/logger.js'
import { NoInstallationFoundError } from '../../exceptions/no-installation-found-error.js'
import { getDependencyTree } from '../npm/get-dependency-tree.js'
import { getDirectoryPrefix } from '../npm/get-directory-prefix.js'
import { getInstalledVersionPaths } from '../npm/get-installed-version-paths.js'
import { getPackageData } from '../npm/get-package-data.js'
import { getPackageTrees } from '../npm/get-package-trees.js'
import { getTreePredecessor } from '../npm/get-tree-predecessor.js'
import { DependencyTree, PackageData } from '../npm/interfaces.js'

/**
 * Finds tracked source files and then filters them based on ones that appear in a  project which
 * depends on the in-context instrumented package/version.
 *
 * @param instrumentedPackage - Data about the instrumented package to use during filtering.
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param fileExtensions - List of file extensions to capture metrics for.
 * @param logger - Logger instance.
 * @returns A (possibly empty) array of source files.
 */
export async function findRelevantSourceFiles(
  instrumentedPackage: PackageData,
  cwd: string,
  root: string,
  fileExtensions: string[],
  logger: Logger
) {
  logger.traceEnter('', 'findRelevantSourceFiles', [instrumentedPackage, cwd, root, fileExtensions])
  const sourceFiles = await getTrackedSourceFiles(root, logger, fileExtensions)

  const dependencyTree = await getDependencyTree(cwd, root, logger)

  const filterPromises = sourceFiles.map(async (f) => {
    const prefix = await getDirectoryPrefix(path.dirname(f.fileName), logger)
    const prefixPackageData = await getPackageData(prefix, root, logger)

    let packageTrees = getPackageTrees(dependencyTree, prefixPackageData)

    let instrumentedInstallVersions: string[] | undefined = undefined
    let shortestPathLength: number | undefined = undefined
    do {
      for (const tree of packageTrees) {
        const instrumentedInstallPaths = getInstalledVersionPaths(tree, instrumentedPackage.name)
        if (instrumentedInstallPaths.length > 0) {
          const pathsLength = instrumentedInstallPaths[0]?.length ?? 0
          if (shortestPathLength === undefined || pathsLength < shortestPathLength) {
            instrumentedInstallVersions = instrumentedInstallPaths.map(
              (path) => getPropertyByPath(tree, path)['version']
            )
            shortestPathLength = pathsLength
          }
        }
      }
      // did not find, go up one level for all packages
      packageTrees = packageTrees
        .map((tree) => getTreePredecessor(dependencyTree, tree['path'] as ObjectPath))
        .filter((tree) => tree !== undefined) as DependencyTree[]
    } while (shortestPathLength === undefined && packageTrees.length > 0)

    if (instrumentedInstallVersions === undefined) {
      throw new NoInstallationFoundError(instrumentedPackage.name)
    }

    return instrumentedInstallVersions.some((version) => version === instrumentedPackage.version)
  })
  const filterData = await Promise.all(filterPromises)

  const results = sourceFiles.filter((_, index) => {
    return filterData[index]
  })

  logger.traceEnter('', 'findRelevantSourceFiles', results)
  return results
}
