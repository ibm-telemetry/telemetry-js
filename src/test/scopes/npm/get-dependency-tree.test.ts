/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'path'
import { describe, expect, it } from 'vitest'

import { getDependencyTree } from '../../../main/scopes/npm/get-dependency-tree.js'
import { Fixture } from '../../__utils/fixture.js'
import { initLogger } from '../../__utils/init-logger.js'

describe('getDependencyTree', () => {
  const logger = initLogger()

  it('correctly obtains dependency tree', async () => {
    const root = new Fixture(path.join('projects', 'complex-nesting-thingy'))
    const cwd = new Fixture(
      path.join('projects', 'complex-nesting-thingy', 'node_modules', 'package1')
    )
    await expect(getDependencyTree(cwd.path, root.path, logger)).resolves.toMatchSnapshot()
  })
})
