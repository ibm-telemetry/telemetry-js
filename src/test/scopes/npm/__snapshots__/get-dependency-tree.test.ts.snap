// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getDependencyTree > correctly obtains dependency tree 1`] = `
{
  "dependencies": {
    "package1": {
      "dependencies": {
        "instrumented": {
          "overridden": false,
          "version": "1.0.0",
        },
      },
      "overridden": false,
      "resolved": "file:../package1",
      "version": "1.0.0",
    },
    "package2": {
      "overridden": false,
      "resolved": "file:../package2",
      "version": "1.0.0",
    },
  },
  "name": "basic-monorepo",
  "version": "1.0.0",
}
`;

exports[`getDependencyTree > correctly obtains dependency tree for nested dep 1`] = `
{
  "dependencies": {
    "intermediate": {
      "dependencies": {
        "deeply-nested": {
          "dependencies": {
            "instrumented": {
              "overridden": false,
              "version": "1.0.0",
            },
          },
          "overridden": false,
          "version": "1.0.0",
        },
      },
      "overridden": false,
      "version": "1.0.0",
    },
  },
  "name": "hoisted-deeply-nested-deps",
  "version": "1.0.0",
}
`;
